indexer: index more details about blocks and transactions

altergui opened the pull request:
needed since we started pruning the blockstore

altergui committed:
indexer: CreateBlock during Commit instead of OnBeginBlock

since the CreateBlock now fetches the whole block, we need app.NodeClient initialized
in all nodes (including seeds), else seeds panic on first Commit

altergui committed:
indexer: index more details about blocks and transactions

blocks:
* indexerdb: add block details chain_id, proposer_address and last_block_hash
* indexerdb: rename method GetBlock -> GetBlockByHeight
* indexerdb: make CreateBlock an UPSERT
* indexer: new method BlockByHeight (and indexerdb.GetBlockByHash)
* indexer: new method SearchBlocks (and indexerdb.SearchBlocks)

transactions:
* indexerdb: add raw_tx, subtype, signature and signer in transactions table
* indexerdb: make CreateTransaction an UPSERT
* indexer: new methods CountTransactionsByHeight and SearchTransactions
* indexer: rename GetTransaction* methods -> GetTxMetadata*

indexer migrations:
 * 0014_alter_columns_table_blocks.sql
 * 0015_alter_columns_table_transactions.sql

indexer: reindex blocks and transactions after migrations
indexer: new method ReindexBlocks

* vochaintx: add tx.TxSubtype method

p4u commented on api/helpers.go:
```diff
@@ -51,6 +52,38 @@ func (a *API) sendTx(tx []byte) (*cometcoretypes.ResultBroadcastTx, error) {
+			Block: comettypes.Block{
+				Header:     tmblock.Header,
+				Data:       tmblock.Data,
+				Evidence:   tmblock.Evidence,
```
Do we need evidence?

altergui replied:
i haven't touched this, it's just the previous code, and i'd leave it as-is.

but better yet, i've simplified this and now just use tmblock altogether

altergui replied:
https://github.com/vocdoni/vocdoni-node/compare/dbdd78ff508f48e60a450f43d5a5d139a50ef990..711f82e200237bdd406173d8d5874eb462b3a96a

mvdan commented on api/chain.go:
```diff
@@ -422,9 +421,16 @@ func (a *API) chainEstimateDateHandler(_ *apirest.APIdata, ctx *httprouter.HTTPC
 		return err
 	}
 	timestamp := a.vocinfo.HeightTime(height)
+	if timestamp == nil {
```
FWIW I would keep HeightTime to returning a non-pointer, and you can use the IsZero method to check if it's valid/present or not.

mvdan commented on api/helpers.go:
```diff
@@ -51,6 +52,33 @@ func (a *API) sendTx(tx []byte) (*cometcoretypes.ResultBroadcastTx, error) {
 	return resp, nil
 }
 
+func (a *API) getBlockFromAppOrIndexer(height int64) (*Block, error) {
```
why is it useful to query either the app or the indexer? wouldn't it be enough to always just query one? if this is necessary at least I would document it

altergui replied:
good point, maybe we could refactor this, but for now i just added a comment

altergui replied:
actually, in the end, i split up the crucial hotfix (#1339) and removed this "getBlockFromAppOrIndexer" completely

altergui committed:
api: fetch blocks and transactions from indexer rather than app BlockStore

these endpoints now fetch blocks from indexer, return just `header` and include `txCount`
 * /chain/blocks
 * /chain/blocks/{height}
 * /chain/blocks/hash/{hash}

this new endpoint fetches the full tx from indexer, and includes `subtype` and `signer` fields
 * /chain/transactions/{hash}

this legacy endpoint is now marked as deprecated
 * /chain/transactions/{height}/{index}

this endpoint now accepts more filter params (subtype, signer)
 * /chain/transactions

refactor:
* api: rename chainBlockHandler -> chainBlockByHeightHandler

altergui committed:
indexer: drop concept of transaction id, hash is now the primary key

 * api: remove endpoint /chain/transactions/reference/index/{index}

indexer migration:
 * 0013_recreate_table_transactions.sql

altergui committed:
api: rename /chain/transactions fields

renames:
  * transactionHash -> hash
  * blockHeight -> height
  * transactionIndex -> index
  * transactionType -> type

affected endpoints:
  /chain/blocks/{height}/transactions/page/{page}
  /chain/transactions/page/{page}
  /chain/transactions
  /chain/transactions/{height}/{index}
  /chain/transactions/reference/{hash}
